### DISCUSSION TITLE
Use without flakes?

### DISCUSSION THREAD

**Discussion by keirlawson** [gmodena/nix-flatpak#104] [Q&A] [answered]
I have a regular NixOS configuration.nix that does not make use of flakes, is it possible to use nix-flatpak in this scenario without flakes? If not, would it be possible to add support for this?

**Comment by gmodena** [comment]
Hey @keirlawson,

`nix-flatpak` uses flakes for ease of distribution, but under the hood it's just a nix module.  You should be able to use it without flakes by manually cloning the repository (e.g. with `builtins.fetchGit` or `fetchTarball`) and importing the module (e.g. `nix-flatpak/modules/nixos.nix`) in your `configuration.nix`.

That's a happy case scenario that I have never tried though.

**Comment by boustanihani** [comment]
@keirlawson have you been successful in using `nix-flatpak` as suggested (without flakes)?

If yes, could you please share your code :) ?

**Comment by gmodena** [comment]
Apologies for the lag on this. I finally managed to sandbox some time to try manual installation.

I'll need to update the doc and I could use some help from users, but
tl;dr: manually downloading nix-flatpak and importing in `configuration.nix` seems to work. The workflow can be automated as follows:
```
# configuration.nix
let
  # Fetch nix-flatpak 
  nix-flatpak = pkgs.fetchFromGitHub {
    owner = "gmodena";
    repo = "nix-flatpak";
    rev = "v0.6.0";
    sha256 = "0s3mpb28rcmma29vv884fi3as926bfszhn7v8n74bpnp5qg5a1c8";
  };
in
  imports = [
    # Import the nix-flatpak module 
    "${nix-flatpak}/modules/nixos.nix"
  ];
  # ... your config

  # Configure nix-flatpak
  services.flatpak = {
    enable = true;
    packages = [
      "org.mozilla.firefox"
    ];
  };
```

Here I'm installing on NixOS; for home-manager you should import `"${nix-flatpak}/modules/home-manager.nix` where appropriate. I'm not familiar with channels (I started using nix with flakes), but I assume that would require to upstream some extra boilerplate to support the API. It's something I'd consider doing if there's enough demand, or if someone would contributed a patch.

For now, I hope that documenting manual install would be a good start.

**Comment by gmodena** [comment, marked as answer]
I've added a section on Manual Installation to the project README doc.

[Here's ](https://gist.github.com/gmodena/535f33651db735e39fdb74f7c24a56ac)a self-contained configuration.nix that you can build and try out in a qemu vm (see comments in file for setup steps).
